feat: add descriptor namespaces

imcelroy opened the pull request:
**Public API Changes**
<!-- Describe any changes to the public API, or write "None" -->
None

**Description**
<!-- Describe what has changed, and motivation behind those changes -->
This will add a new namespace called `descriptor` for each ROS package interface generated by `generate-ros-messages`. The new namespace is found inside the msg|srv|action namespace for each package. For example:

```typescript
namespace builtin_interfaces {
    namespace msg {
      export interface Duration {
        sec: number;
        nanosec: number;
      }
      export interface DurationConstructor {
        new(other?: Duration): Duration;
      }
      export interface Time {
        sec: number;
        nanosec: number;
      }
      export interface TimeConstructor {
        new(other?: Time): Time;
      }
      namespace descriptor {
        export interface Duration {
          sec: 'int32';
          nanosec: 'uint32';
        }
        export interface Time {
          sec: 'int32';
          nanosec: 'uint32';
        }
      }
    }
  }
``` 
The descriptor interfaces always have the format `{field: "<interface_type>"}`. For example:

```typescript
  namespace geometry_msgs {
    namespace msg {
      namespace descriptor {
        export interface PointStamped {
          header: 'std_msgs/msg/Header';
          point: 'geometry_msgs/msg/Point';
        }
      }
    } 
  }  
```
<!-- Link relevant GitHub issues -->
Here is the discussion: [discussions/1091](https://github.com/RobotWebTools/rclnodejs/discussions/1091)

imcelroy committed:
feat: add descriptor namespaces

imcelroy commented:
Hi @minggangw, I would like to add a couple unit tests, but I'm not sure where/how you test the content of the `interfaces.d.ts` file? If you could point me in the right direction I'll add the unit tests and then open this for review!

imcelroy committed:
fix saveActionAsTSD

minggangw commented:
Hi @imcelroy thanks for submitting the PR quickly, be honest, currently, we don't have kind of unit test targeting on the generated messages for typescript specifically.

We have https://github.com/RobotWebTools/rclnodejs/blame/develop/test/types/index.test-d.ts that leverages [dst](https://github.com/tsdjs/tsd) to test the type definitions, I think we can add some test cases that verify the generated types under `descriptor` namespace, we don't need to test each, but one or two at least, how do you think?

imcelroy commented:
Hi @minggangw, yes that sounds good to me! I will take a look and add those asap.

imcelroy committed:
fix add Constructor interfaces in description namespace

imcelroy committed:
add descriptor interface tsd tests

imcelroy commented:
@minggangw I have added the unit tests and opened this for review, thanks for your help!

imcelroy committed:
use expectAssignable

imcelroy commented:
@minggangw It's strange, `npx tsd` is passing for me locally, I'm not sure why it isn't passing in the CI. I'm looking into this, but if you have any ideas let me know.

imcelroy committed:
use Fibonacci action as example for unit test

imcelroy commented:
All seems to be good now! The Windows build that is failing seems to be present on other pull requests as well, and not directly linked to my changes. Will that block us from merging this feature?

minggangw commented:
> All seems to be good now! The Windows build that is failing seems to be present on other pull requests as well, and not directly linked to my changes. Will that block us from merging this feature?

I suspect the failure on Windows platform may be caused by env changes, because all tests do pass on ROS2 jazzy build. I will start the review work soon, thanks!

minggangw commented on rostsd_gen/index.js:
```diff
@@ -238,16 +223,84 @@ function savePkgInfoAsTSD(pkgInfos, fd) {
+}
+
+function saveMsgAsTSD(rosMsgInterface, fd, descriptorInterfaceType = false) {
+  const indentlevel = descriptorInterfaceType ? 8 : 6;
```
nit: I prefer renaming `indentlevel` because it confused with `indentLevel` on L.292

minggangw commented on rostsd_gen/index.js:
```diff
@@ -270,7 +324,8 @@ function saveMsgFieldsAsTSD(
   typePrefix = '',
-  useSamePackageSubFolder = false
+  useSamePackageSubFolder = false,
+  descriptorInterfaceType = false
```
I prefer renaming `descriptorInterfaceType` to `willGenerateDescriptionInterface`

minggangw commented on test/types/index.test-d.ts:
```diff
@@ -390,3 +390,40 @@ param.value.integer_value = BigInt(123);
+);
+expectAssignable<'action_msgs/msg/GoalInfo'>(
+  cancelGoalRequestDescriptor.goal_info
+);
```
really like these ut 👍

imcelroy replied:
Thanks!

minggangw commented on rostsd_gen/index.js:
```diff
@@ -119,47 +121,30 @@ function savePkgInfoAsTSD(pkgInfos, fd) {
+
+      // generate descriptor msg/srv/action interfaces
+      fs.writeSync(fd, `      namespace ${descriptorInterfaceNamespace} {\n`);
+      const descriptorInterfaceType = true;
```
`descriptorInterfaceType` => `willGenerateDescriptionInterface`

minggangw commented on rostsd_gen/index.js:
```diff
@@ -238,16 +223,84 @@ function savePkgInfoAsTSD(pkgInfos, fd) {
     isInternalServiceEventMsgInterface(rosMsgInterface);
-  saveMsgFieldsAsTSD(rosMsgInterface, fd, 8, ';', '', useSamePkg);
-  fs.writeSync(fd, '      }\n');
+  const indentLevel = descriptorInterfaceType ? 10 : 8;
```
Is it possible to abstract the functionality of getting the value of indent to a separate function in this file?

minggangw reviewed (commented):
Thanks for submitting the PR, the implementation looks good and some comments left.

minggangw commented:
You can add your name into https://github.com/RobotWebTools/rclnodejs/blob/develop/CONTRIBUTORS.md, it's up to you 😄

imcelroy committed:
fix review comments + linting

- rename willGenerateDescriptorInterface
- create getIndentSpacing funciton
- avoid similar indentLevel variable names

imcelroy committed:
separate contributors, add new contributions

imcelroy commented:
@minggangw Thank you for the review! I think I have addressed all of your comments, let me know what you think!

minggangw reviewed (approved):
LGTM!

minggangw commented:
@imcelroy I have merged the PR, if you want to land the change in a new release asap, let me know, I can prepare for it this week.

imcelroy commented:
@minggangw Thank you very much! Yes, if you could add this to a new release asap that would be great! I appreciate it!

minggangw commented:
@imcelroy https://www.npmjs.com/package/rclnodejs/v/0.33.0 is online :tada:

imcelroy commented:
@minggangw wow! Thank you very much for the speed; that was impressive! 🎉
